test/coordinatesystems.jl

```diff
@@ -0,0 +1,435 @@
+                     rθ_gn.θ.partials.data[1] rθ_gn.θ.partials.data[2] ]
+        m = transform_deriv(p_from_c, xy)
+        @test m ≈ m_gn
+
+        rθ_gn = Polar(GradientNumber(2.23606797749979, (1.0, 0.0)), GradientNumber(-2.0344439357957027, (0.0, 1.0)))
+        xy_gn = transform(c_from_p, rθ_gn)
+        m_gn = @fsa [xy_gn[1].partials.data[1] xy_gn[1].partials.data[2];
+                     xy_gn[2].partials.data[1] xy_gn[2].partials.data[2] ]
+        m = transform_deriv(c_from_p, rθ)
+        @test m ≈ m_gn
+    end
+
+    @testset "3D" begin
+        s_from_cart = SphericalFromCartesian()
+        cart_from_s = CartesianFromSpherical()
+        cyl_from_cart = CylindricalFromCartesian()
+        cart_from_cyl = CartesianFromCylindrical()
+        cyl_from_s = CylindricalFromSpherical()
+        s_from_cyl = SphericalFromCylindrical()
+        identity_cart = IdentityTransformation{Point{3}}()
+        identity_s = IdentityTransformation{Spherical}()
+        identity_cyl = IdentityTransformation{Cylindrical}()
+
+        # inverses
+        @test inv(s_from_cart) == cart_from_s
+        @test inv(cart_from_s) == s_from_cart
+        @test inv(cyl_from_cart) == cart_from_cyl
+        @test inv(cart_from_cyl) == cyl_from_cart
+        @test inv(s_from_cyl) == cyl_from_s
+        @test inv(cyl_from_s) == s_from_cyl
+
+        # composition of inverses
+        @test s_from_cart ∘ cart_from_s == identity_s
+        @test cart_from_s ∘ s_from_cart == identity_cart
+        @test cyl_from_cart ∘ cart_from_cyl == identity_cyl
+        @test cart_from_cyl ∘ cyl_from_cart == identity_cart
+        @test s_from_cyl ∘ cyl_from_s == identity_s
+        @test cyl_from_s ∘ s_from_cyl == identity_cyl
+
+        # cyclic composition
+        @test s_from_cart ∘ cart_from_cyl == s_from_cyl
+        @test cart_from_s ∘ s_from_cyl == cart_from_cyl
+        @test cyl_from_cart ∘ cart_from_s == cyl_from_s
+        @test cart_from_cyl ∘ cyl_from_s == cart_from_s
+        @test s_from_cyl ∘ cyl_from_cart == s_from_cart
+        @test cyl_from_s ∘ s_from_cart == cyl_from_cart
+
+        # Spherical <-> Cartesian
+        # test all 8 octants of the sphere (for consistency of branch-cuts)
+
+        # Octant 1
+        xyz = Point(1.0, 2.0, 3.0)
+        rθϕ = Spherical(3.7416573867739413, 1.1071487177940904, 0.9302740141154721)
+        @test transform(s_from_cart, xyz) ≈ rθϕ
+        @test transform(cart_from_s, rθϕ) ≈ xyz
+
+        xyz_gn = Point(GradientNumber(1.0, (1.0, 0.0, 0.0)), GradientNumber(2.0, (0.0, 1.0, 0.0)), GradientNumber(3.0, (0.0, 0.0, 1.0)))
+        rθϕ_gn = transform(s_from_cart, xyz_gn)
+        m_gn = @fsa [rθϕ_gn.r.partials.data[1] rθϕ_gn.r.partials.data[2] rθϕ_gn.r.partials.data[3];
+                     rθϕ_gn.θ.partials.data[1] rθϕ_gn.θ.partials.data[2] rθϕ_gn.θ.partials.data[3];
+                     rθϕ_gn.ϕ.partials.data[1] rθϕ_gn.ϕ.partials.data[2] rθϕ_gn.ϕ.partials.data[3] ]
```

Comment:
Sure, though I'd just go with `GradientNumber` for now just as you have here, but cleaning things up so that it looks like

```julia
xyz_gn = make_dual(Point, 1.0, 2.0, 3.0)
rθϕ_gn = transform(s_from_cart, xyz_gn)
m_gn = extract_partials(rθϕ_gn)
```

In this form it doesn't matter so much what upstream breakage happens, since you can simply fix `make_dual` and `extract_partials` (the names may be bad, not sure what they should be).